tests/Unit/DataStructures/DataBox/Test_DataBox.cpp

```diff
@@ -2684,6 +2684,37 @@ void test_serialization() noexcept {
   serialization_compute_items_of_base_tags();
   serialization_of_pointers();
 }
+
+void test_get_mutable_reference() noexcept {
+  INFO("test get_mutable_reference");
+  // Make sure the presence of tags that could not be extracted
+  // doesn't prevent the function from working on other tags.
+  auto box = db::create<db::AddSimpleTags<test_databox_tags::Tag0,
+                                          test_databox_tags::Tag2, Parent<0>>,
+                        db::AddComputeTags<test_databox_tags::Tag4Compute>>(
+      3.14, "Original string"s,
+      std::make_pair(Boxed<int>(std::make_shared<int>(5)),
+                     Boxed<double>(std::make_shared<double>(3.5))));
+
+  decltype(auto) ref =
+      db::get_mutable_reference<test_databox_tags::Tag2>(make_not_null(&box));
+  static_assert(std::is_same_v<decltype(ref), std::string&>);
+  decltype(auto) base_ref =
+      db::get_mutable_reference<test_databox_tags::Tag2Base>(
+          make_not_null(&box));
+  static_assert(std::is_same_v<decltype(base_ref), std::string&>);
+  CHECK(&ref == &base_ref);
+  ref = "New string";
+  CHECK(db::get<test_databox_tags::Tag2Base>(box) == "New string");
+
+  // These should all fail to compile:
+  // db::get_mutable_reference<test_databox_tags::Tag0>(make_not_null(&box));
```

Comment:
Can we use SFINAE to check this explicitly? Maybe it's too much hassle/compile time overhead, but I'm imagining something that uses `void_t<decltype(/*invalid_function_call*/)>` with a template specialization.
If I am not mistaken, it is permissible to define template structs inside functions.

Reply:
SFINAE only applies to function signatures and similar.  Problems in the bodies are always errors.  (And inside decltype it will only look at the body if it actually needs to, although in this case it would because of the auto return type.)

Reply:
Understood, all good then.